server/test/write_authzmodel.go

```diff
@@ -362,28 +491,78 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
 					},
 				},
 			},
-			err: errors.RelationNotFound("owner", "", nil),
+			err: errors.CannotAllowMultipleReferencesToOneRelation,
 		},
 		{
-			name: "ExecuteWriteFailsIfUnknownRelationInTupleToUsersetComputedUserset",
+			name: "ExecuteWriteFailsIfIntersectionIncludesSameRelationTwice",
 			request: &openfgapb.WriteAuthorizationModelRequest{
-				TypeDefinitions: &openfgapb.TypeDefinitions{
-					TypeDefinitions: []*openfgapb.TypeDefinition{
-						{
-							Type: "repo",
-							Relations: map[string]*openfgapb.Userset{
-								"writer": {Userset: &openfgapb.Userset_This{}},
-								"viewer": {
-									Userset: &openfgapb.Userset_TupleToUserset{
-										TupleToUserset: &openfgapb.TupleToUserset{
-											Tupleset: &openfgapb.ObjectRelation{
-												Object:   "",
-												Relation: "writer",
-											},
+				StoreId: storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": {
+								Userset: &openfgapb.Userset_Intersection{
+									Intersection: &openfgapb.Usersets{Child: []*openfgapb.Userset{
+										{Userset: &openfgapb.Userset_ComputedUserset{
 											ComputedUserset: &openfgapb.ObjectRelation{
-												Object:   "$TUPLE_USERSET_OBJECT",
-												Relation: "owner",
+												Relation: "viewer",
 											},
+										}},
+										{Userset: &openfgapb.Userset_This{}},
+									}},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: errors.CannotAllowMultipleReferencesToOneRelation,
+		},
+	}
+
+	ctx := context.Background()
+	logger := logger.NewNoopLogger()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cmd := commands.NewWriteAuthorizationModelCommand(datastore, logger)
+			resp, err := cmd.Execute(ctx, test.request)
+			require.ErrorIs(t, err, test.err)
+
+			if err == nil {
+				require.True(t, id.IsValid(resp.AuthorizationModelId))
+			}
+		})
+	}
+}
+
+func TestWriteAuthorizationModelWithRelationalTypes(t *testing.T, datastore storage.OpenFGADatastore) {
+	var tests = []struct {
+		name    string
+		request *openfgapb.WriteAuthorizationModelRequest
+		err     error
+	}{
+		{
+			name: "succeeds on a valid model with a type type",
```

Comment:
with a type type?

Reply:
Yeah. haha. I'll update it to read "with a objectType type".